src/stores/CallStore.ts

```diff
@@ -0,0 +1,188 @@
+import WidgetStore from "./WidgetStore";
+import SettingsStore from "../settings/SettingsStore";
+import { SettingLevel } from "../settings/SettingLevel";
+import MediaDeviceHandler, { MediaDeviceKindEnum } from "../MediaDeviceHandler";
+import { Call, CallEvent, ConnectionState } from "../models/Call";
+
+export enum CallStoreEvent {
+    // Signals a change in the call associated with a given room
+    Call = "call",
+    // Signals a change in the active call
+    ActiveCall = "active_call",
+}
+
+export class CallStore extends AsyncStoreWithClient<null> {
+    private static _instance: CallStore;
+    public static get instance(): CallStore {
+        if (!this._instance) {
+            this._instance = new CallStore();
+            this._instance.start();
+        }
+        return this._instance;
+    }
+
+    private constructor() {
+        super(defaultDispatcher);
+    }
+
+    protected async onAction(payload: ActionPayload): Promise<void> {
+        // nothing to do
+    }
+
+    protected async onReady(): Promise<any> {
+        // We assume that the calls present in a room are a function of room
+        // state and room widgets, so we initialize the room map here and then
+        // update it whenever those change
+        for (const room of this.matrixClient.getRooms()) {
+            this.updateRoom(room);
+        }
+        this.matrixClient.on(ClientEvent.Room, this.onRoom);
+        this.matrixClient.on(RoomStateEvent.Events, this.onRoomState);
+        WidgetStore.instance.on(UPDATE_EVENT, this.onWidgets);
+
+        // If the room ID of the last connected call is still in settings at
+        // this time, that's a sign that we failed to disconnect from it
+        // properly, and need to clean up after ourselves
+        const uncleanlyDisconnectedRoomId = SettingsStore.getValue("activeCallRoomId");
+        if (uncleanlyDisconnectedRoomId) {
+            await this.get(uncleanlyDisconnectedRoomId)?.clean();
+            SettingsStore.setValue("activeCallRoomId", null, SettingLevel.DEVICE, null);
+        }
+    }
+
+    protected async onNotReady(): Promise<any> {
+        for (const call of this.calls.values()) call.destroy();
+        this.calls.clear();
+        this.matrixClient.off(ClientEvent.Room, this.onRoom);
+        this.matrixClient.off(RoomStateEvent.Events, this.onRoomState);
+        WidgetStore.instance.off(UPDATE_EVENT, this.onWidgets);
+    }
+
+    private _activeCall: Call | null = null;
+    /**
+     * The call to which the user is currently connected.
+     */
+    public get activeCall(): Call | null {
+        return this._activeCall;
+    }
+    private set activeCall(value: Call | null) {
+        this._activeCall = value;
+        this.emit(CallStoreEvent.ActiveCall, value);
+
+        // The room ID is persisted to settings so we can detect unclean disconnects
+        SettingsStore.setValue("activeCallRoomId", null, SettingLevel.DEVICE, value?.roomId ?? null);
```

Comment:
I sort of agree, but this is what Travis had to say back when I added this flag to the original video rooms implementation:

> Temporary flags like this should go through the settings framework in the event we ever decide to support devices which don't have localstorage (a thing that's unlikely to happen, but likely enough to mention/defend for it)

It is also nice that the settings framework abstracts away the details of localstorage, so you don't have to worry about the fact that localstorage holds a string rather than the raw boolean.